src/commands/run.rs

```diff
@@ -370,12 +396,49 @@ impl RunCommand {
         result
     }
 
-    fn load_main_module(
-        &self,
-        store: &mut Store<Host>,
-        linker: &mut Linker<Host>,
-        module: Module,
-    ) -> Result<()> {
+    fn setup_epoch_handler(&self, store: &mut Store<Host>, module_name: &str) -> Box<dyn FnOnce()> {
+        if let Some(path) = &self.profile_guest {
+            let interval = self.profile_guest_interval;
+            let profiler = Arc::new(Mutex::new(GuestProfiler::new(module_name, interval)));
+
+            let cloned = profiler.clone();
+            if let Some(timeout) = self.wasm_timeout {
+                let mut timeout = (timeout.as_secs_f64() / interval.as_secs_f64()).ceil() as u64;
+                assert!(timeout > 0);
+                store.epoch_deadline_callback(move |store| {
+                    cloned.lock().unwrap().sample(store);
+                    timeout -= 1;
+                    if timeout == 0 {
+                        bail!("timeout exceeded");
+                    }
+                    Ok(1)
+                });
+            } else {
+                store.epoch_deadline_callback(move |store| {
+                    cloned.lock().unwrap().sample(store);
+                    Ok(1)
+                });
+            }
+
+            store.set_epoch_deadline(1);
+            let engine = store.engine().clone();
+            thread::spawn(move || loop {
+                thread::sleep(interval);
+                engine.increment_epoch();
+            });
+
+            let path = path.clone();
+            return Box::new(move || {
+                let mut profiler = profiler.lock().unwrap();
+                if let Err(e) = std::fs::File::create(&path)
+                    .map_err(anyhow::Error::new)
+                    .and_then(|output| profiler.finish(std::io::BufWriter::new(output)))
+                {
+                    eprintln!("failed writing profile at {}: {e:#}", path.display());
+                }
+            });
```

Comment:
Doesn't matter a ton in this case, but one pattern I've used to make APIs like this slightly harder to get wrong is to return an `impl Drop`:

```rust
struct FinishEpochHandler<F: FnOnce()>(Option<F>);
impl<F: FnOnce()> Drop for FinishEpochHandler<F> {
    fn drop(&mut self) {
        if let Some(f) = self.0.take() {
            f();
        }
    }
}
return FinishEpochHandler(Some(move || {
    // what you have now...
}));
```

This way, callers don't need to actually remember to call the resulting function, at most they just need to add a scope.

But I guess if you do Alex's suggestion of moving the profiler into `Store`, then this result function would require a store argument, and the RAII/`impl Drop` approach won't really work anymore...